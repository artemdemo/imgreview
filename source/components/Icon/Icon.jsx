import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import './Icon.less';

/**
 * Icon component
 * @param props
 * @link https://fontawesome.com/get-started
 */
const Icon = (props) => {
    const { name, inText, className, title } = props;

    if (name === '' || name == null) {
        throw new Error('Icon prop `name` couldn\'t be empty');
    }

    const iconClass = classnames(className, 'fa', {
        [`fa-${name}`]: true,
        'icon_in-text': inText,
    });
    return (
        <span
            className={iconClass}
            title={title}
        />
    );
};

Icon.propTypes = {
    name: PropTypes.string.isRequired,
    className: PropTypes.string,
    title: PropTypes.string,
    inText: PropTypes.bool,
};

Icon.defaultProps = {
    className: '',
    title: null,
    inText: false,
};

export default Icon;
