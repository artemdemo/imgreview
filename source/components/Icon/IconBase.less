@import '~font-awesome/css/font-awesome.min.css';

.icon_in-text {
    padding-right: 5px;
}
