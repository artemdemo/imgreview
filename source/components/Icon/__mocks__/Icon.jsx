import React from 'react';

const Icon = props => (
    <div data-mock='Icon'>
        {props.children}
    </div>
);

export default Icon;
