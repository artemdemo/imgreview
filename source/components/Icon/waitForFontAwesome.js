import Promise from 'bluebird';

/**
 * Check that font awesome has been loaded
 * @type {Promise<any>}
 * @link https://stackoverflow.com/a/24618517
 */
const waitForFontAwesome = new Promise((resolve, reject) => {
    let retries = 50;

    const checkReady = function() {
        retries -= 1;
        const canvas = document.createElement('canvas');
        canvas.width = 20;
        canvas.height = 20;
        const context = canvas.getContext('2d');
        context.fillStyle = 'rgba(0,0,0,1.0)';
        context.fillRect(0, 0, 20, 20);
        context.font = '16pt FontAwesome';
        context.textAlign = 'center';
        context.fillStyle = 'rgba(255,255,255,1.0)';
        context.fillText('\uf0c8', 10, 18);
        const { data } = context.getImageData(2, 10, 1, 1);
        if (data[0] !== 255 && data[1] !== 255 && data[2] !== 255) {
            if (retries > 0) {
                setTimeout(checkReady, 300);
            } else {
                reject();
            }
        } else {
            resolve();
        }
    };

    checkReady();
});

export default waitForFontAwesome;
