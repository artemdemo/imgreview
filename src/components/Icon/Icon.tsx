import React from 'react';
import classnames from 'classnames';
import styled from 'styled-components';
import withIconFont from './withIconFont';

import 'font-awesome/css/font-awesome.min.css';

export type TIconProps = {
    name: string;
    className?: string;
    title?: string;
    inText?: boolean;
};

type TState = {
    fontLoaded: boolean;
};

const IconSty = styled.span<{inText?: boolean}>`
    padding-right: ${props => props.inText ? '5px' : '0'};
`;

/**
 * Icon component
 * @param props
 * @link https://fontawesome.com/get-started
 */
class Icon extends React.PureComponent<TIconProps, TState> {
    static readonly defaultProps = {
        className: '',
        title: null,
        inText: false,
    };

    render() {
        const { name, inText, className, title } = this.props;

        if (name === '' || name == null) {
            throw new Error('Icon prop `name` couldn\'t be empty');
        }

        const iconClass = classnames(className, 'fa', {
            [`fa-${name}`]: true,
        });

        return (
            <IconSty
                className={iconClass}
                title={title}
                inText={inText}
            />
        );
    }
}

export default withIconFont(Icon);
