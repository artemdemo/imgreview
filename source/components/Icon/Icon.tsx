import React from 'react';
import { TIconProps } from './IconBase';

/**
 * Notice: This file has `.tsx` extension.
 * I need it only to support JSX code in mock file.
 * Typescript doesn't allow to write jsx code in `.ts` files, only in `.tsx`
 */
import lazify from '../../containers/lazify';

const LazyIcon = lazify(() => import(/* webpackChunkName: "<IconBase />" */ './IconBase'));

const IconPlaceholder = (props: TIconProps) => {
    return (
        <LazyIcon
            {...props}
        >
            {props.title}
        </LazyIcon>
    );
};

export default IconPlaceholder;
