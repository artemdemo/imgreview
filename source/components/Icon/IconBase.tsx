import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import waitForFontAwesome from './waitForFontAwesome';

import './IconBase.less';

export type TIconProps = {
    name: string;
    className?: string;
    title?: string;
    inText?: boolean;
};

type State = {
    fontLoaded: boolean;
};

/**
 * Icon component
 * @param props
 * @link https://fontawesome.com/get-started
 */
class Icon extends React.PureComponent<TIconProps, State> {
    private fontAwesomePromise: Promise<any>;
    private unmounted: boolean = false;

    static readonly defaultProps = {
        className: '',
        title: null,
        inText: false,
    };

    constructor(props) {
        super(props);

        this.state = {
            fontLoaded: false,
        };
    }

    componentDidMount() {
        this.fontAwesomePromise = waitForFontAwesome
            .then(() => {
                if (!this.unmounted) {
                    this.setState({
                        fontLoaded: true,
                    });
                }
            })
            .catch(() => {
                console.warn('Error while loading font awesome');
            });
    }

    componentWillUnmount() {
        this.unmounted = true;
    }

    render() {
        const { name, inText, className, title } = this.props;

        if (name === '' || name == null) {
            throw new Error('Icon prop `name` couldn\'t be empty');
        }

        const iconClass = classnames(className, 'fa', {
            [`fa-${name}`]: true,
            'icon_in-text': inText,
        });

        if (this.state.fontLoaded) {
            return (
                <span
                    className={iconClass}
                    title={title}
                />
            );
        }

        return (
            <span>{title}</span>
        );
    }
}

export default Icon;
