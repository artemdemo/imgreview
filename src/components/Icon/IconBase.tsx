import React from 'react';
import classnames from 'classnames';
import styled from 'styled-components';
import waitForFontAwesome from './waitForFontAwesome';

import 'font-awesome/css/font-awesome.min.css';

export type TIconProps = {
    name: string;
    className?: string;
    title?: string;
    inText?: boolean;
};

type State = {
    fontLoaded: boolean;
};

const IconSty = styled.span<{inText?: boolean}>`
    padding-right: ${props => props.inText ? '5px' : '0'};
`;

/**
 * Icon component
 * @param props
 * @link https://fontawesome.com/get-started
 */
class Icon extends React.PureComponent<TIconProps, State> {
    private fontAwesomePromise: Promise<any>;
    private unmounted: boolean = false;

    static readonly defaultProps = {
        className: '',
        title: null,
        inText: false,
    };

    constructor(props) {
        super(props);

        this.state = {
            fontLoaded: false,
        };
    }

    componentDidMount() {
        this.fontAwesomePromise = waitForFontAwesome
            .then(() => {
                if (!this.unmounted) {
                    this.setState({
                        fontLoaded: true,
                    });
                }
            })
            .catch(() => {
                console.warn('Error while loading font awesome');
            });
    }

    componentWillUnmount() {
        this.unmounted = true;
    }

    render() {
        const { name, inText, className, title } = this.props;

        if (name === '' || name == null) {
            throw new Error('Icon prop `name` couldn\'t be empty');
        }

        const iconClass = classnames(className, 'fa', {
            [`fa-${name}`]: true,
        });

        if (this.state.fontLoaded) {
            return (
                <IconSty
                    className={iconClass}
                    title={title}
                    inText={inText}
                />
            );
        }

        return (
            <IconSty
                inText={inText}
            >
                {title}
            </IconSty>
        );
    }
}

export default Icon;
