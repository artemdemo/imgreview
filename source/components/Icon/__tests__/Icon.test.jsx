/* eslint-disable new-cap */
import React from 'react';
import renderer from 'react-test-renderer';
import Icon from '../Icon';

describe('Icon', () => {
    it('should render with name', () => {
        const tree = renderer.create(
            <Icon name='fire' />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render with name and className', () => {
        const tree = renderer.create(
            <Icon name='fire'className='some-class' />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('shoudl render with name and inText', () => {
        const tree = renderer.create(
            <Icon name='fire' inText />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should throw an error if no name provided', () => {
        expect(() => {
            Icon({ name: '' });
        }).toThrow();

        expect(() => {
            Icon({});
        }).toThrow();
    });
});
