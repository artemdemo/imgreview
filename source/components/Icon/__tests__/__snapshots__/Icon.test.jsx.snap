// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon shoudl render with name and inText 1`] = `
<span
  className="fa fa-fire icon_in-text"
  title={null}
/>
`;

exports[`Icon should render with name 1`] = `
<span
  className="fa fa-fire"
  title={null}
/>
`;

exports[`Icon should render with name and className 1`] = `
<span
  className="some-class fa fa-fire"
  title={null}
/>
`;

exports[`Icon should render with name and title 1`] = `
<span
  className="fa fa-fire"
  title="Fire"
/>
`;
